`Cache::get` Incorrectly Wraps `CancellationException` in `CompletionException`

tl:dr `get` handles `CancellationException` incorrectly, triggering unwanted Caffeine warning logs

# Problem

The current implementation of `Cache::get` uses a manual logic to catch exceptions thrown in the provided `compute` function, then re-throw it in the completable future using `thenApply` (emphasis below mine):

```kotlin
   suspend fun get(key: K, compute: suspend (K) -> V): V {
      val scope = CoroutineScope(coroutineContext)
      var error: Throwable? = null
      val value = cache.get(key) { k, _ ->
         val asCompletableFuture = scope.async {
            try {
               compute(k)
            } catch (e: Throwable) {
               error = e
               null
            }
         }.asCompletableFuture()

         // Here ------------------------------v
         asCompletableFuture.thenApply { it ?: throw error ?: NullPointerException() }
      }.await()
      error?.let { throw it }
      return value
   }
```

While this does work, exceptions like `CancellationException` and `TimeoutException` are then wrapped in `CompletionException` as stated in `CompletionStage`'s [documentation](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html):

> In all other cases, if a stage's computation terminates abruptly with an (unchecked) exception or error, then all dependent stages requiring its completion complete exceptionally as well, with a [CompletionException](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionException.html) holding the exception as its cause.

[Caffeine logs a WARN](https://github.com/ben-manes/caffeine/blob/5cc036327d4caa5b620a70b98f5d4ed337abf3d3/caffeine/src/main/java/com/github/benmanes/caffeine/cache/LocalAsyncCache.java#L207) when exceptions are thrown except for `CancellationException` and `TimeoutException`, so this wrapping of `CancellationException` results in an unnecessary log:

```java
  default void handleCompletion(...) {
    valueFuture.whenComplete((value, error) -> {
      long loadTime = cache().statsTicker().read() - startTime;
      if (value == null) {
        if ((error != null) && !(error instanceof CancellationException)
            && !(error instanceof TimeoutException)) {
          logger.log(Level.WARNING, "Exception thrown during asynchronous load", error);
        }
        ...
      } else ...
```

# Impact

Other than the logging happening when the `compute` coroutine is cancelled, some libraries such as Arrow's `Raise` utilise the `CancellationException` to work with structured concurrency. For instance, my team wanted to create a helper function that works with Arrow's `Either`, caching only when the value is `Either.Right`:

```
suspend fun <K : Any, V : Any, L> Cache<K, V>.get(
    key: K,
    mappingFunction: suspend (K) -> Either<L, V>
): Either<L, V> =
    either { get(key) { k -> mappingFunction(k).bind() } }
```

The `bind` method throws a subtype of `CancellationException` under the hood, so this results in Caffeine warning about the exception thrown:

```
Jan 29, 2026 11:17:56 AM com.github.benmanes.caffeine.cache.LocalAsyncCache lambda$handleCompletion$0
WARNING: Exception thrown during asynchronous load
java.util.concurrent.CompletionException: arrow.core.raise.RaiseCancellationExceptionNoTrace: kotlin.coroutines.cancellation.CancellationException should never get cancelled. Always re-throw it if captured.This swallows the exception of Arrow's Raise, and leads to unexpected behavior.When working with Arrow prefer Either.catch or arrow.core.raise.catch to automatically rethrow CancellationException.
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
    ...
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: arrow.core.raise.RaiseCancellationExceptionNoTrace: kotlin.coroutines.cancellation.CancellationException should never get cancelled. Always re-throw it if captured.This swallows the exception of Arrow's Raise, and leads to unexpected behavior.When working with Arrow prefer Either.catch or arrow.core.raise.catch to automatically rethrow CancellationException.
```

Of course, since exceptions are rethrown unwrapped by the `await()` call, the code still works. We could manually disable the logging from the Cache, but in my opinion a better solution would be to correctly handle the exceptions in the library.

# Solution

The issue stems from using the parent's coroutine scope to create `async` child coroutines:

```kotlin
   // Current implementation
   suspend fun get(key: K, compute: suspend (K) -> V): V {
      val scope = CoroutineScope(coroutineContext)
      var error: Throwable? = null
      val value = cache.get(key) { k, _ ->
         val asCompletableFuture = scope.async { ... }.asCompletableFuture()
         ...
      }.await()
      error?.let { throw it }
      return value
   }
```

When the child coroutine throws a non-cancellation exception, it results in the parent coroutine immediately terminating and throwing the same exception as well. Past patches have addressed this by manually catching and rethrowing the exception later.

Instead of manually catching and throwing exceptions, we can create a new `coroutineScope` to re-raise any exceptions thrown within the `compute` method:

```kotlin
   suspend fun get(key: K, compute: suspend (K) -> V): V =
      coroutineScope { cache.get(key) { k, _ -> async { compute(k) }.asCompletableFuture() }.await() }
```

Which can be simplified further down using the `future` builder

```kotlin
   suspend fun get(key: K, compute: suspend (K) -> V): V =
      coroutineScope { cache.get(key) { k, _ -> future { compute(k) } }.await() }
```

When `compute` throws a `CancellationException`, it will gracefully cancel the `CompletableFuture` as well without being wrapped in a `CompletionException`. When it throws a non-cancellation exception, the outer `coroutineScope` will throw instead of outright cancelling the parent coroutine.

--- a/aedile-core/src/main/kotlin/com/sksamuel/aedile/core/Cache.kt
+++ b/aedile-core/src/main/kotlin/com/sksamuel/aedile/core/Cache.kt
@@ -4,11 +4,12 @@ import com.github.benmanes.caffeine.cache.AsyncCache
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Deferred
 import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
 import kotlinx.coroutines.future.asCompletableFuture
 import kotlinx.coroutines.future.asDeferred
 import kotlinx.coroutines.future.await
+import kotlinx.coroutines.future.future
 import java.util.concurrent.CompletableFuture
-import java.util.concurrent.Executor
 import kotlin.coroutines.coroutineContext
 
 class Cache<K : Any, V : Any>(
@@ -52,25 +53,8 @@ class Cache<K : Any, V : Any>(
     * @return the present value, the computed value, or throws.
     *
     */
-   suspend fun get(key: K, compute: suspend (K) -> V): V {
-      val scope = CoroutineScope(coroutineContext)
-      var error: Throwable? = null
-      val value = cache.get(key) { k, _ ->
-         val asCompletableFuture = scope.async {
-            // if compute throws, then it will cause the parent coroutine to be cancelled as well
-            // we don't want that, as want to throw the exception back to the caller.
-            // so we must capture it and throw it manually
-            try {
-               compute(k)
-            } catch (e: Throwable) {
-               error = e
-               null
-            }
-         }.asCompletableFuture()
-         asCompletableFuture.thenApply { it ?: throw error ?: NullPointerException() }
-      }.await()
-      error?.let { throw it }
-      return value
+   suspend fun get(key: K, compute: suspend (K) -> V): V = coroutineScope {
+      cache.get(key) { k, _ -> future { compute(k) } }.await()
    }
 
    /**
@@ -140,24 +124,8 @@ class Cache<K : Any, V : Any>(
       return cache.asMap().mapValues { it.value.asDeferred() }
    }
 
-   suspend fun getAll(keys: Collection<K>, compute: suspend (Collection<K>) -> Map<K, V>): Map<K, V> {
-      val scope = CoroutineScope(coroutineContext)
-      var error: Throwable? = null
-      val value = cache.getAll(keys) { ks: Set<K>, _: Executor ->
-         scope.async {
-            // if compute throws, then it will cause the parent coroutine to be cancelled as well
-            // we don't want that, as want to throw the exception back to the caller.
-            // so we must capture it and throw it manually
-            try {
-               compute(ks)
-            } catch (e: Throwable) {
-               error = e
-               emptyMap()
-            }
-         }.asCompletableFuture()
-      }.await()
-      error?.let { throw it }
-      return value
+   suspend fun getAll(keys: Collection<K>, compute: suspend (Collection<K>) -> Map<K, V>): Map<K, V> = coroutineScope {
+      cache.getAll(keys) { ks, _ -> future { compute(ks) } }.await()
    }
 
    /**

--- a/aedile-core/src/main/kotlin/com/sksamuel/aedile/core/LoadingCache.kt
+++ b/aedile-core/src/main/kotlin/com/sksamuel/aedile/core/LoadingCache.kt
@@ -5,9 +5,11 @@ import com.github.benmanes.caffeine.cache.AsyncLoadingCache
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Deferred
 import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
 import kotlinx.coroutines.future.asCompletableFuture
 import kotlinx.coroutines.future.asDeferred
 import kotlinx.coroutines.future.await
+import kotlinx.coroutines.future.future
 import java.util.concurrent.CompletableFuture
 import kotlin.coroutines.coroutineContext
 
@@ -66,11 +68,11 @@ class LoadingCache<K : Any, V>(
     *
     * See full docs at [AsyncCache.getAll].
     */
-   suspend fun getAll(keys: Collection<K>, compute: suspend (Set<K>) -> Map<K, V & Any>): Map<K, V & Any> {
-      val scope = CoroutineScope(coroutineContext)
-      @Suppress("UNCHECKED_CAST") // getAll returns CompletableFuture<Map<K, @NonNull V>>
-      return cache.getAll(keys) { k, _ -> scope.async { compute(k.toSet()) }.asCompletableFuture() }.await() as Map<K, V & Any>
-   }
+   suspend fun getAll(keys: Collection<K>, compute: suspend (Set<K>) -> Map<K, V & Any>): Map<K, V & Any> =
+      coroutineScope {
+         @Suppress("UNCHECKED_CAST") // getAll returns CompletableFuture<Map<K, @NonNull V>>
+         cache.getAll(keys) { ks, _ -> future { compute(ks) } }.await() as Map<K, V & Any>
+      }
 
    /**
     * Returns the value associated with a key in this cache, getting that value from the
@@ -86,26 +88,8 @@ class LoadingCache<K : Any, V>(
     *
     * See full docs at [AsyncLoadingCache.get].
     */
-   suspend fun get(key: K, compute: suspend (K) -> V): V {
-
-      val scope = CoroutineScope(coroutineContext)
-      var error: Throwable? = null
-      val value = cache.get(key) { k, _ ->
-         val asCompletableFuture = scope.async {
-            // if compute throws, then it will cause the parent coroutine to be cancelled as well
-            // we don't want that, as want to throw the exception back to the caller.
-            // so we must capture it and throw it manually
-            try {
-               compute(k)
-            } catch (e: Throwable) {
-               error = e
-               null
-            }
-         }.asCompletableFuture()
-         asCompletableFuture.thenApply { it ?: throw error ?: NullPointerException() }
-      }.await()
-      error?.let { throw it }
-      return value
+   suspend fun get(key: K, compute: suspend (K) -> V): V = coroutineScope {
+      cache.get(key) { k, _ -> future { compute(k) } }.await()
    }
 
    /**

